### DISCUSSION TITLE
asyncComputed as promise

### DISCUSSION THREAD

**Discussion by bpolaszek** [vueuse/vueuse#1216] [Ideas]
Hello there,

First of all, thank you all for your brilliant work! 🎉 
VueUse is a real life saver and makes Vue adoption a breeze.

I'd like to submit an idea, which is something like `asyncComputed(callback, initialState, {promise: true}` so that in this case, instead of returning the `computed` directly, return a _Promise_ resolved by the `computed`.
The use case is to use it along with top-level `await` with `<script setup>` and `Suspense` components.

Example usage:
```vue
<script setup>
const car = await store.dispatch('getItem', '/api/cars/1'); // {'@id': "/api/cars/1", driver: "/api/employees/1", name: "Company's Tesla"}
const driver = await asyncComputed(
    async () => await store.dispatch('getItem', car.driver),
    undefined,
    {promise: true}
); // {'@id': "/api/employees/1", name: "Anthony Fu"}

// At some point, `car.driver` changes to `/api/employees/2`
// Then `driver` is automatically recomputed as `{'@id': "/api/employees/2", name: "Yasser Lahbibi"}` 
// - by fetching that item through the API or getting it already fetched from the store
</script>
```

I didn't succeed in doing it by myself (TS is not my primary language as I'm more a back-end guy 😅) so I wrote a kinda crappy wrapper which does the job, but it would be great if `asyncComputed()` could do this out of the box with an opt-in option:
```js
const computedAsPromise = async (callback, defaultValue) => {
  const evaluating = ref(false);
  const value = asyncComputed(callback, defaultValue, {evaluating});
  return new Promise(resolve => {
    watch(evaluating, (isEvaluating) => {
      if (false === isEvaluating) {
        resolve(value);
      }
    });
  });
};
```

What do you think?

**Comment by yassilah** [comment]
Hi! I'm just wondering, couldn't you simply write: 

```vue
<script setup>
const book = await store.dispatch('getItem', '/api/books/1')
// book is defined now so you can refer to it and fetch your author without any computed, can't you?
const author = await store.dispatch('getItem', book.author)
</script>
```

**Reply by bpolaszek** [reply]
Hi!

Because, what if `book.author` changes? Then `author` isn't recomputed. 

I could use `watch` for this, that's pretty much what I used to do before, but `asyncComputed` seems to be the most appropriate one-liner, except it's not usable in a top-level `await` context, as it won't block until the value is resolved.

Maybe "book/author" example isn't the perfect one as a book's author actually rarely changes, but this just illustrates the mechanism of a relation between 2 items that is represented through an url (what Hydra APIs are made of), and that relation is subject to changes.

I just updated the original example with a "car/driver" relation so that it looks more relevant.

**Reply by yassilah** [reply]
Hey! Sorry yes I understand what you mean, I just thought since `book` or `car` weren't reactive, there wasn't going to be any changes. But of course if the `author` or `driver` properties are in fact reactive, your example makes sense 🙂

**Reply by bpolaszek** [reply]
Yes, forgot to mention that! 😅  Whatever item I get from the store is actually reactive.
